### ISSUE TITLE
infix operators for `isequal` and `isless`

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#21908]
Given the demise of https://github.com/JuliaLang/julia/issues/19148, the current state of affairs might be somewhat more palatable if `isequal` and `isless` had operator forms. One option would be making `eq` and `lt` infix operators like `isa` and `in`, but I don't love that idea. The operators I proposed in #19148 could be used: `~==`, `~!=`, `~<`, `~>`, `~<=` and `~>=`. Or some Unicode operators.

**Comment by yuyichao** [comment]
Making it easier to work with FP values as normal bits sounds good. As long as they don't use the operators that people normally expect to have IEEE complaint behaviors.

**Comment by yuyichao** [comment]
As for the name, `~==` is a bit too similar to MATLAB's `~=` ...

**Comment by JeffBezanson** [comment]
It seems quite likely the best place to go from here is just to allow infix for all identifiers. The main issue is probably precedence. Many of these (including the existing `in` and `isa`) would use comparison precedence, and even chained comparison (`a isless b isless c`), but I'm not sure how to handle it in general. Maybe a fixed list of comparisons, and give everything else the same precedence level (perhaps equal to `|>`?).

**Comment by JeffBezanson** [comment]
Of course there are some useful unicode characters like `≅` and `≨`.

**Comment by ararslan** [comment]
I don't see the benefit of using these functions infix rather than prefix. It would seem especially unfortunate to me to allow infix for all identifiers.

**Comment by TotalVerb** [comment]
Infix for identifiers would mean some trouble for macros and matrices, unless infix identifiers are disallowed in space-sensitive context (which is possibly reasonable, but not the way things currently work).

**Comment by TotalVerb** [comment]
(I would even go far as to suggest that infix `in` and `isa` should be **removed**, but I suppose the ship on that has sailed.)

**Comment by yurivish** [comment]
Simple juxtaposition was considered a binary operator in Fortress. I wonder if something like that idea could work for Julia?

**Comment by StefanKarpinski** [comment]
Disallowing infix operators in space-sensitive context would make sense. I personally use infix `in` all the time and enjoy `isa` when I have occasion to use it.

**Comment by ararslan** [comment]
I was the one who added infix `isa`, but I'll echo Jeff's concerns at the time that it's a rather slippery slope in terms of allowing infix things like this. I would argue that `in` and `isa` are very special cases and their infix use shouldn't generalize to just any operator. IMO we should generally keep words as prefix and Unicode/ASCII symbols as infix. Having Unicode operators to express non-IEEE operations seems like a reasonable compromise between convenience and consistency.

**Comment by oscardssmith** [comment]
We could go the java route with [equals](https://docs.oracle.com/javase/8/docs/api/java/lang/Float.html#equals-java.lang.Object-), where equals compares `nan`s equal and 0s unequal, while `==` does the opposite, but that sounds like a recipe for badness.

**Comment by ararslan** [comment]
@oscardssmith See [`isequal`](https://docs.julialang.org/en/latest/stdlib/base.html#Base.isequal-Tuple{Any,Any})

**Comment by StefanKarpinski** [comment]
@oscardssmith: that was essentially my proposal in https://github.com/JuliaLang/julia/issues/19148.

**Comment by StefanKarpinski** [comment]
Since I have these saved in a note and not recorded on GitHub, I should post the set of infix Unicode operators that I found that could be used for this which all have a `~` above the normal comparison operators:

- `≅ = isequal`
- `≇ = !isequal`
- `⪝ = isless`
- `⪞ = (x, y) -> isless(y, x)`
- `⪟ = (x, y) -> !isless(y, x)`
- `⪠ = !isless`

I don't think there's any other style of Unicode operator that covers the infix forms you'd want.

**Comment by JeffreySarnoff** [comment]
I see the vertical symmetry in your choices.
Allow reflection and the symbology simplifies.  

| | | | |
|----------|:-----------:|:----------:|----------|
| `isequal` | **≅** |   **≆** | `!isequal`  |
| | | | |
| `isless` | **⪅** |   **⪉** | `!isless` |
 | | | | |

**Comment by brenhinkeller** [comment]
It seems like this could be done without breaking anything right? And if you don't want to use the new operators you wouldn't have to, so no reason we couldn't still consider this in 1.x if so..

**Comment by StefanKarpinski** [comment]
Yes, if someone is already using these in their own code, the way `using` works will leave their code intact. The only issue that can happen is if some package is exporting them and some other code uses them without explicitly importing, but that's a potential issue whenever Base exports a new name.
